File: client/config/guestConfig.ts
/**
 * Guest Session Configuration
 * Centralized configuration for guest mode following DRY principles
 * Industry-grade security and session management
 */

// ============================================================================
// GUEST SESSION CONSTANTS (Single Source of Truth)
// ============================================================================

/**
 * Maximum number of prompts a guest user can send
 * Industry standard: 10-20 prompts for trial/guest access
 */
export const GUEST_PROMPT_LIMIT = 15;

/**
 * Guest session expiration time in milliseconds
 * 24 hours = 86,400,000 ms
 * Industry standard: 24-48 hours for guest sessions
 */
export const GUEST_SESSION_EXPIRY_MS = 24 * 60 * 60 * 1000;

/**
 * AsyncStorage key for guest session
 * Prefixed with @ for React Native convention
 */
export const GUEST_SESSION_STORAGE_KEY = '@guest_session';

/**
 * Guest session ID prefix for identification
 */
export const GUEST_SESSION_ID_PREFIX = 'guest_';

// ============================================================================
// SECURITY CONFIGURATION
// ============================================================================

/**
 * Session validation rules
 */
export const GUEST_SESSION_VALIDATION = {
  /**
   * Minimum session ID length (timestamp + random)
   * Prevents session ID guessing attacks
   */
  MIN_SESSION_ID_LENGTH: 20,
  
  /**
   * Maximum prompt count (safety check)
   * Prevents tampering with local storage
   */
  MAX_PROMPT_COUNT: GUEST_PROMPT_LIMIT,
  
  /**
   * Session expiry buffer (5 minutes)
   * Prevents edge cases with clock skew
   */
  EXPIRY_BUFFER_MS: 5 * 60 * 1000,
} as const;

// ============================================================================
// HELPER FUNCTIONS (DRY Principle)
// ============================================================================

/**
 * Generate a secure guest session ID
 * Format: guest_<timestamp>_<random>
 * 
 * @returns Unique session ID
 */
export const generateGuestSessionId = (): string => {
  const timestamp = Date.now();
  const random = Math.random().toString(36).substring(2, 15);
  return `${GUEST_SESSION_ID_PREFIX}${timestamp}_${random}`;
};

/**
 * Check if a session is expired
 * Includes buffer time for clock skew
 * 
 * @param expiresAt - Session expiration timestamp
 * @returns True if session is expired
 */
export const isSessionExpired = (expiresAt: number): boolean => {
  return Date.now() >= (expiresAt - GUEST_SESSION_VALIDATION.EXPIRY_BUFFER_MS);
};

/**
 * Validate session data integrity
 * Prevents tampering with local storage
 * 
 * @param session - Session data to validate
 * @returns True if session is valid
 */
export const validateGuestSession = (session: any): boolean => {
  // Check required fields
  if (!session || typeof session !== 'object') {
    return false;
  }

  if (!session.id || typeof session.id !== 'string') {
    return false;
  }

  if (!session.id.startsWith(GUEST_SESSION_ID_PREFIX)) {
    return false;
  }

  if (session.id.length < GUEST_SESSION_VALIDATION.MIN_SESSION_ID_LENGTH) {
    return false;
  }

  // Validate prompt count
  if (typeof session.promptCount !== 'number' || session.promptCount < 0) {
    return false;
  }

  if (session.promptCount > GUEST_SESSION_VALIDATION.MAX_PROMPT_COUNT) {
    console.warn('⚠️ Guest session prompt count exceeds limit, possible tampering');
    return false;
  }

  // Validate timestamps
  if (typeof session.createdAt !== 'number' || session.createdAt <= 0) {
    return false;
  }

  if (typeof session.expiresAt !== 'number' || session.expiresAt <= 0) {
    return false;
  }

  // Check if expiry is reasonable (not more than 48 hours from creation)
  const maxExpiry = session.createdAt + (48 * 60 * 60 * 1000);
  if (session.expiresAt > maxExpiry) {
    console.warn('⚠️ Guest session expiry is unreasonable, possible tampering');
    return false;
  }

  return true;
};

/**
 * Calculate remaining prompts
 * 
 * @param promptCount - Current prompt count
 * @returns Number of prompts remaining
 */
export const calculateRemainingPrompts = (promptCount: number): number => {
  return Math.max(0, GUEST_PROMPT_LIMIT - promptCount);
};

/**
 * Check if prompt limit is reached
 * 
 * @param promptCount - Current prompt count
 * @returns True if limit is reached
 */
export const isPromptLimitReached = (promptCount: number): boolean => {
  return promptCount >= GUEST_PROMPT_LIMIT;
};

// ============================================================================
// DISCLAIMER TEXT (Centralized)
// ============================================================================

/**
 * Guest mode disclaimer text
 * Used across the app for consistency
 */
export const GUEST_DISCLAIMER = {
  /**
   * Short disclaimer for banners/tooltips
   */
  SHORT: `Guest mode: ${GUEST_PROMPT_LIMIT} prompts available`,
  
  /**
   * Full disclaimer for modals/pages
   */
  FULL: `You are using AI.ttorney in guest mode. Guest accounts are limited to ${GUEST_PROMPT_LIMIT} prompts and expire after 24 hours. Sign up for unlimited access and to save your chat history.`,
  
  /**
   * Call-to-action text
   */
  CTA: 'Sign up for unlimited access',
  
  /**
   * Limit reached message
   */
  LIMIT_REACHED: `You've reached the ${GUEST_PROMPT_LIMIT}-prompt limit for guest mode. Please sign up to continue using AI.ttorney.`,
  
  /**
   * Session expired message
   */
  SESSION_EXPIRED: 'Your guest session has expired. Please sign in or continue as a new guest.',
} as const;

// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

/**
 * Guest session interface
 * Enforces type safety across the app
 */
export interface GuestSession {
  id: string;
  promptCount: number;
  createdAt: number;
  expiresAt: number;
}

/**
 * Guest session validation result
 */
export interface SessionValidationResult {
  isValid: boolean;
  reason?: string;
}

// ============================================================================
// EXPORT ALL
// ============================================================================

export default {
  GUEST_PROMPT_LIMIT,
  GUEST_SESSION_EXPIRY_MS,
  GUEST_SESSION_STORAGE_KEY,
  GUEST_SESSION_ID_PREFIX,
  GUEST_SESSION_VALIDATION,
  GUEST_DISCLAIMER,
  generateGuestSessionId,
  isSessionExpired,
  validateGuestSession,
  calculateRemainingPrompts,
  isPromptLimitReached,
};

File: server/config/guest_config.py
"""
Guest Configuration - Backend
Single source of truth for guest rate limiting configuration
Mirrors frontend config for consistency (DRY principle)
"""

import os
from typing import Final

# ============================================================================
# GUEST SESSION CONSTANTS (Single Source of Truth)
# ============================================================================

# Maximum number of prompts a guest user can send
# Must match frontend: client/config/guestConfig.ts
GUEST_PROMPT_LIMIT: Final[int] = int(os.getenv("GUEST_PROMPT_LIMIT", "15"))

# Guest session expiration time in hours
# Must match frontend: 24 hours
SESSION_EXPIRY_HOURS: Final[int] = int(os.getenv("GUEST_SESSION_HOURS", "24"))

# IP-based rate limiting (backup layer)
# Prevents abuse even if session tokens are compromised
IP_RATE_LIMIT_PER_HOUR: Final[int] = int(os.getenv("GUEST_IP_LIMIT_HOUR", "30"))

# ============================================================================
# VALIDATION RULES
# ============================================================================

class GuestSessionValidation:
    """
    Guest session validation rules
    Mirrors frontend validation for consistency
    """
    
    # Minimum session ID length (prevents guessing attacks)
    MIN_SESSION_ID_LENGTH: Final[int] = 20
    
    # Session ID must start with this prefix
    SESSION_ID_PREFIX: Final[str] = "guest_"
    
    # Maximum session age in seconds (24 hours + buffer)
    MAX_SESSION_AGE_SECONDS: Final[int] = (SESSION_EXPIRY_HOURS + 1) * 3600
    
    # Expiry buffer for clock skew (5 minutes)
    EXPIRY_BUFFER_SECONDS: Final[int] = 5 * 60

# ============================================================================
# ERROR MESSAGES (Centralized)
# ============================================================================

class GuestErrorMessages:
    """
    Centralized error messages for guest rate limiting
    Ensures consistent user-facing messages
    """
    
    @staticmethod
    def limit_reached(prompts: int, hours_left: int, minutes_left: int) -> str:
        """Generate limit reached message"""
        return f"You've used all {prompts} free prompts. Resets in {hours_left}h {minutes_left}m."
    
    @staticmethod
    def session_expired() -> str:
        """Generate session expired message"""
        return "Your 24-hour session has expired. Refresh to start a new session."
    
    @staticmethod
    def session_not_found() -> str:
        """Generate session not found message"""
        return "Session expired. Please refresh the page."
    
    @staticmethod
    def ip_rate_limit(minutes_left: int) -> str:
        """Generate IP rate limit message"""
        return f"Too many requests from your IP. Try again in {minutes_left} minutes."

# ============================================================================
# HELPER FUNCTIONS (DRY Principle)
# ============================================================================

def calculate_remaining_prompts(prompt_count: int) -> int:
    """
    Calculate remaining prompts
    
    Args:
        prompt_count: Current prompt count
        
    Returns:
        Number of prompts remaining
    """
    return max(0, GUEST_PROMPT_LIMIT - prompt_count)

def is_prompt_limit_reached(prompt_count: int) -> bool:
    """
    Check if prompt limit is reached
    
    Args:
        prompt_count: Current prompt count
        
    Returns:
        True if limit is reached
    """
    return prompt_count >= GUEST_PROMPT_LIMIT

def validate_session_id(session_id: str) -> bool:
    """
    Validate session ID format
    
    Args:
        session_id: Session ID to validate
        
    Returns:
        True if session ID is valid
    """
    if not session_id or not isinstance(session_id, str):
        return False
    
    if not session_id.startswith(GuestSessionValidation.SESSION_ID_PREFIX):
        return False
    
    if len(session_id) < GuestSessionValidation.MIN_SESSION_ID_LENGTH:
        return False
    
    return True

def calculate_reset_time(created_at: float) -> int:
    """
    Calculate when session resets
    
    Args:
        created_at: Session creation timestamp
        
    Returns:
        Reset timestamp
    """
    return int(created_at + SESSION_EXPIRY_HOURS * 3600)

# ============================================================================
# EXPORT ALL
# ============================================================================

__all__ = [
    'GUEST_PROMPT_LIMIT',
    'SESSION_EXPIRY_HOURS',
    'IP_RATE_LIMIT_PER_HOUR',
    'GuestSessionValidation',
    'GuestErrorMessages',
    'calculate_remaining_prompts',
    'is_prompt_limit_reached',
    'validate_session_id',
    'calculate_reset_time',
]
